### PR TITLE
next: refactor transaction management

### PR DESCRIPTION
**Pull request by Steven-Chan**
connect #682

There is a safeXXXStore for each resource that interact with database, which would panic if function called without transaction.

### CODE REVIEW

File: pkg/auth/dependency/provider/anonymous/provider.go [hunk 1]
```diff
@@ -11,14 +11,18 @@ type providerImpl struct {
...
+func newProvider(builder db.SQLBuilder, executor db.SQLExecutor, logger *logrus.Entry) *providerImpl {
 	return &providerImpl{
 		sqlBuilder:  builder,
 		sqlExecutor: executor,
 		logger:      logger,
 	}
 }
 
+func NewProvider(builder db.SQLBuilder, executor db.SQLExecutor, logger *logrus.Entry) Provider {
+	return NewProvider(builder, executor, logger)
```
Comment (carmenlau) [inline pkg/auth/dependency/provider/anonymous/provider.go]: s/NewProvider/newProvider/

File: pkg/core/skyerr/recover.go [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 import (
+	"fmt"
 	"runtime/debug"
 
 	"github.com/sirupsen/logrus"
 	"github.com/skygeario/skygear-server/pkg/server/skyerr"
 )
 
 func ErrorFromRecoveringPanic(r interface{}) skyerr.Error {
+	fmt.Printf("%+v\n", r)
```
Comment (carmenlau) [inline pkg/core/skyerr/recover.go]: Debug print?

File: pkg/record/handler/save.go [hunk 3]
```diff
@@ -522,10 +525,44 @@ func RecordSaveHandler(req *RecordModifyRequest, resp *RecordModifyResponse) (er
...
+			return do()
+		})
+		if txErr != nil {
+			err = skyerr.MakeError(txErr)
+		}
+
+		return
+	}
+
+	err = do()
```
Comment (carmenlau) [inline pkg/record/handler/save.go]: Should we also wrap this error with `skyerr.MakeError`? So the behaviour of atomic and non-atomic is the same

**Comment by Steven-Chan** [comment]
> Should we also wrap this error with `skyerr.MakeError`? So the behaviour of atomic and non-atomic is the same

@carmenlau  I have updated the code to remove `skyerr.MakeError` when inatomic, because APIHandler does not require the handle function to return a skygear error, it is the responsibility of the transformation function to make it skygear error.

**Pull request merged**